File: Search a 2D Matrix - Leetcode 74/Search a 2D Matrix - Leetcode 74.js
var searchMatrix = function(matrix, target) {
    let m = matrix.length;
    let n = matrix[0].length;
    let l = 0;
    let r = m * n - 1;

    while (l <= r) {
        let mid = Math.floor((l + r) / 2);
        let i = Math.floor(mid / n);
        let j = mid % n;
        let midNum = matrix[i][j];

        if (target === midNum) {
            return true;
        } else if (target < midNum) {
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }

    return false;
};

File: Search a 2D Matrix - Leetcode 74/Search a 2D Matrix - Leetcode 74.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();
        int l = 0;
        int r = m * n - 1;

        while (l <= r) {
            int mid = (l + r) / 2;
            int i = mid / n;
            int j = mid % n;
            int midNum = matrix[i][j];

            if (target == midNum) {
                return true;
            } else if (target < midNum) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        return false;
    }
};
